### ISSUE TITLE
Meteor 1.6 -> Permission denied

### ISSUE THREAD

**Issue by gitTerebi** [zodern/meteor-up#787]
I'm getting permission problems using newer meteord image. Looks like tar doesn't have permissions to extract files?

App was built on Windows 10.

If I try it on my OSX box it builds and deploys without any problems.

```
mup.cmd deploy
Building App Bundle Locally

WARNING: The output directory is under your source tree.
         Your generated files may get interpreted as source code!
         Consider building into a different directory instead
         meteor build ../output


Started TaskList: Pushing Meteor App
[172.81.177.153] - Pushing Meteor App Bundle to The Server
[172.81.177.153] - Pushing Meteor App Bundle to The Server: SUCCESS
[172.81.177.153] - Prepare Bundle
[172.81.177.153] x Prepare Bundle: FAILED
	
	-----------------------------------STDERR-----------------------------------
	/es5-ext/test/array/#/exclusion.js: Cannot open: Permission denied
	tar: bundle/programs/server/npm/node_modules/meteor/matb33_collection-hooks/node_modules/es5-ext/test/array/#/fill: Cannot mkdir: Permission denied
	tar: bundle/programs/server/npm/node_modules/meteor/matb33_collection-hooks/node_modules/es5-ext/test/array/#/filter: Cannot mkdir: Permission denied
	tar: bundle/programs/server/npm/node_modules/meteor/matb33_collection-hooks/node_modules/es5-ext/test/array/#/find: Cannot mkdir: Permission denied
	tar: bundle/programs/server/npm/node_modules/meteor/matb33_collection-hooks/node_modules/es5-ext/test/array/#/find-index: Cannot mkdir: Permission denied
	tar: bundle/programs/server/npm/node_modules/meteor/matb33_collection-hooks/node_modules/es5-ext/test/array/#/first-index.js: Cannot open: Permission denied
	tar: bundle/programs/server/npm/node_modules/meteor/matb33_collection-hooks/node_modules/es5-ext/test/array/#/first.js: Cannot open: Permission denied
	tar: bundle/programs/serv
	-----------------------------------STDOUT-----------------------------------
	node-8.4.0-base: Pulling from abernix/meteord
	Digest: sha256:04f44cf999da383d11de77027ced8b70818222c9890a405ce10803af3ee5a642
	Status: Image is up to date for abernix/meteord:node-8.4.0-base
	----------------------------------------------------------------------------
```

**Comment by n8ores** [comment]
I get the same error (works on OSX but not Win10) but with basic.js instead of exclusion.js. 

Are you deploying to AWS on ubuntu by any chance?

**Comment by gitTerebi** [comment]
Actually if you turn on --verbose you can see that the error is for all files in the tar. Its just a matter of which one it fails on last depending on what packages.

Uploading to a normal ubuntu image on a vps hosted by luna node. So its not host dependent. 

It works on Mac OSX, just not on win10.

**Comment by n8ores** [comment]
Yes I can see that, I have configured mup.js to login using the user ubuntu (not root) I wonder if the ubuntu user needs to have user permissions added, e.g sudo chown -R /DIRECTORYHOLDING the build.

However I am not sure where the directory that holds the build is located?

**Comment by hope-11** [comment]
I get the same error

**Comment by nisaus** [comment]
I get the same error as well. 

I'm on Windows 10 deploying to AWS on ubuntu. Mine is a different set of js files but as gitTerebi said they are all tar files. I tried setting up and deploying on a clean server but the same error came up. The result is the same for both meteord-base and the 8.4.0 version.

**Comment by hope-11** [comment]
I fix it!
The problem is caused by the permission of user 'ubuntu', so we must set the 'username' 'root' in mup.js.
but first, you should allow user 'root' to be able to login the aws ec2 ubuntu server, by default it's not allowed.
follow these steps:
1. login your server with user 'ubuntu'
2. create root password

> $ sudo passwd root
3. switch to root 

> $ su
4. open the file '/etc/ssh/sshd_config', and edit it

> #vim /etc/ssh/sshd_config
5. modify these words

> PermitRootLogin yes
> PasswordAuthentication yes
6. restart sshd service

> #service sshd restart
7. open the file '/root/.ssh/authorized_keys, and edit it

> #vim /root/.ssh/authorized_keys
8. delete the words before 'ssh-rsa ...', then like this
`ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCPqJ2U4gidqek
4FPQJABENkrUiLVP61LObdFAZXvs2EpCf/nBQCRg4ykMNg+8TC9lb7jC65zfIrTUcNcwongDb4
3k6miSKu1M8fdqXDpcb8CdDRaKpM2wP8l+hTaJ2aWycXmGJ7lZKQPiwNUOhbrOLNEtDmOI9eiV
lz7See98LVLW+6AwfzNA8Cu4riDTvEMQr/WQ9NLrS3BZE1TAAswJi9lGDfTgEvfh4Ji+eI/xT
Xrjkkwjerkjk3jrkwejrkjwe9wASXob4rbV12TXjQIcMKaRGQAGrwOHu0nM2ibfTdgqjrTAG
03CXKzQhF09LdxKlT7GpYe0oVU2R1kjkejwQp howtoing.com`

now, we can login the server with root, please try!

**Comment by gitTerebi** [comment]
The above works!

If on 14.04 you need to do 

> #service ssh restart

Additionally, I didn't need to touch the authorized_keys file. All you need to do is enable root to login over SSH.

**Comment by hope-11** [comment]
I'm glad it works!
I'm not sure why I must modify the authorized_keys file, or it will not work.
Maybe the reason is my vision 16.04.

**Comment by gitTerebi** [comment]
The question is why it works on OSX but not on Windows without having to enable root. Maybe there needs to be a sudo in sh scripts?

**Comment by robport** [comment]
Surely, we should be able to change the permissions of the original user?  Seems odd to force everyone to use root and permit root logins?

**Comment by gitTerebi** [comment]
Agreed, it undermines the entire point of ubuntu instead of root for security reasons.

**Comment by nisaus** [comment]
Thanks for the solution!

**Comment by megaslimegod** [comment]
I tried this solution, but now I get this other error.  

https://github.com/zodern/meteor-up/issues/790

**Comment by nisaus** [comment]
Did anyone figure out what happened regarding the sudden need to enable root access? It's nice that the problem is solved for now but I wouldn't really want to have to use root from here on and out. Thanks!

**Comment by nchokoev** [comment]
Is there a solution without giving up on security?
I'm using it to deploy on AWS, ssh with pem file. 'sudo' doesn't need a pass. Still same issue.

**Comment by zodern** [comment]
I merged @nickich PR. I will release an update tomorrow with the fix.

The last update to mup changed the library used to archive the bundle to fix some of the tar files being corrupted. The new library is probably giving different permissions to the files.

**Comment by gitTerebi** [comment]
Awesome news

**Comment by nisaus** [comment]
That's great, thank you!

**Comment by gjr-qm** [comment]
I tried this change on my Windows 10 machine (deploying a Meteor 1.5.1 app with Meteor up 1.3.5 using abernix/meteord:base docker image and a non-root sudo user for deployment) and it's not quite working for me yet...

First failure i got was `bash: line 28: Dockerfile: Permission denied`. This seems to be caused by the permission on the bundle and can be solved by inserting 'sudo chown ${USER} bundle -R' as line 26 of the file 'plugins/meteor/assets/prepare-bundle.sh'.

Then the process coninues building the image, but when the resulting image starts, I get the following error:

```
tar (child): *.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
```

Still examining, but getting a little tired after 2 days of trying to get something to work that worked flawless before I updated to Meteor 1.6 (using Meteor up 1.2.11). Downgrading to 1.2.11 gave other errors, so right now I'm stuck and can't deploy my application (lost the working version in the process).

**Comment by nchokoev** [comment]
@gjr-qm Where you try to deploy it to? What is the server?

**Comment by gjr-qm** [comment]
Deploying it to my own Ubuntu 16.04 server

**Comment by nchokoev** [comment]
@gjr-qm  Just tested on AWS Ubuntu 16.04 and looks ok. Unfortunately I've no enough info to reproduce your particular case. If you give me more info so I can try to reproduce it... 

BTW, for Meteor 1.6, please try 
```
    docker: {
      image: 'abernix/meteord:node-8.4.0-base',
    },
```

**Comment by gjr-qm** [comment]
@nickich : I also tried that, but that gave me more errors. I thought that was logical, because my application is a Meteor 1.5.1 build.

Strange thing is that it worked perfectly before I updated my computer to Meteor 1.6. Meteor is downloading (and using) version 1.5.1 when I run the project. I thought it would build a 1.5.1 bundle also, which should work with Meteor up 1.3.5 and abernix/meteord:base. Or am I wrong?

Too bad I cannot do a clean Meteor 1.5.1. install anymore (or is there a way to do that?)...

**Comment by gjr-qm** [comment]
@nickich : are you deploying with root user?

**Comment by nchokoev** [comment]
@gjr-qm I'm deploying with 'ubuntu' user (which can 'sudo' without a pass).
As of the first part - it should. Have you tried to upgrade your app to Meteor 1.6 (meteor update --release 1.6) and use the image docker above?

**Comment by zodern** [comment]
I published mup 1.3.6-beta1. Please install it with `npm i -g mup@next` and let me know if the problem is fixed.

**Comment by zodern** [comment]
1.3.6 has been published.

**Comment by gjr-qm** [comment]
With version 1.3.6 the permission problem is solved. Thanx!

After that I still had the problem that the container kept rebooting. Found out that was caused by environment var 'REBUILD_NPM_MODULES=1' I set during my trial&error process based on the fact that bcrypt seemed to have something to do with it and info in the meteord repo. You should not set this environment variable! Just leaving this here in case someone else stumbles upon that error...

My first test container is running not, back to the real work! Thanks again @zodern!

**Comment by wildhart** [comment]
I still get a similar error with 1.3.6 and 1.3.6-beta1.  Trying a fresh AWS Ubuntu 16.04.3 LTS server.  `mup setup` works OK. then trying to deploy gives:

```
$mup deploy --verbose --cached-build

   ┌───────────────────────────────────────────┐
   │                                           │
   │   update available 1.3.6-beta1 => 1.3.6   │
   │        To update, run npm i -g mup        │
   │                                           │
   └───────────────────────────────────────────┘

Not building app. Using build from previous deploy at
/tmp/mup-meteor-86b8821f-c588-4137-875d-ef23740add09

Started TaskList: Pushing Meteor App
[s.virtualinout.com] - Pushing Meteor App Bundle to the Server
[s.virtualinout.com] - Pushing Meteor App Bundle to the Server: SUCCESS
[s.virtualinout.com] - Prepare Bundle
node-8.9.1-base: Pulling from abernix/meteord
c20dc0c5afeb: Already exists
... 
a3ed95caeb02: Already exists
Digest: sha256:9a12326012a41b5ddd2210d5e5d590bb3506bc341a7f5e1540b451e975bb559c
Status: Image is up to date for abernix/meteord:node-8.9.1-base
Error response from daemon: No such container: virtualinout
bash: line 28: Dockerfile: Permission denied
[s.virtualinout.com] x Prepare Bundle: FAILED
	
	-----------------------------------STDERR-----------------------------------
	Error response from daemon: No such container: virtualinout
	bash: line 28: Dockerfile: Permission denied
	-----------------------------------STDOUT-----------------------------------
	node-8.9.1-base: Pulling from abernix/meteord
	c20dc0c5afeb: Already exists
	... 
	a3ed95caeb02: Already exists
	Digest: sha256:9a12326012a41b5ddd2210d5e5d590bb3506bc341a7f5e1540b451e975bb559c
	Status: Image is up to date for abernix/meteord:node-8.9.1-base
	----------------------------------------------------------------------------
```
I'm installing as user ubuntu, not root.  I've also tried abernix/meteord:node-8.4.0-base

I don't know if it's relevant, but when I install/update mup I always get `npm WARN optional dep failed, continuing fsevents@1.1.3`

**Comment by n8ores** [comment]
I also get the Dockerfile: Permission denied error during the mup deploy process. I am running mup deploy FROM my actual AWS server (logged in as ubuntu). 

I am using mup 1.3.6

`[52.64.191.5] - Prepare Bundle
node-8.4.0-base: Pulling from abernix/meteord
Digest: sha256:04f44cf999da383d11de77027ced8b70818222c9890a405ce10803af3ee5a642
Status: Image is up to date for abernix/meteord:node-8.4.0-base
<APP_NAME>
bash: line 28: Dockerfile: Permission denied
[52.64.191.5] x Prepare Bundle: FAILED

        -----------------------------------STDERR-----------------------------------
        bash: line 28: Dockerfile: Permission denied
        -----------------------------------STDOUT-----------------------------------
        node-8.4.0-base: Pulling from abernix/meteord
        Digest: sha256:04f44cf999da383d11de77027ced8b70818222c9890a405ce10803af3ee5a642
        Status: Image is up to date for abernix/meteord:node-8.4.0-base
        <APP_NAME>
        ----------------------------------------------------------------------------
`

**Comment by nchokoev** [comment]
@wildhart, @n8ores 
Just to be clear - did you try removing all running docker containers and all docker images on the host?

**Comment by AdansBatista** [comment]
I added a patch to my app version in 1.5.4 and I'm getting the permission error as well. 
First I tried to add the user to the server on the Docker group because even the command "docker ps" was denied to that user.
Mup version 1.3.7

Now mup is not getting permission to copy files into:

`DEBUG=mup*` mup deploy --cached-build --verbose
  mup:updates checking for updates +0ms8Z4nTKanO2F-y0bVlA.lVzFVlSnCLwRUBqa2zt6RXf29Std4HekcObswe4Y2e0',
  mup:module:default exec => mup deploy +0msir1:m0V9MuWG3d73NAXYvuZu83sg@smtp.sendgrid.net:2525',
  mup:module:meteor exec => mup meteor deploy +0ms27017/MIR-uat',
  mup:module:meteor exec => mup meteor push +4ms
Not building app. Using build from previous deploy at
/tmp/mup-meteor-fe995a1b-f94b-4c4b-9276-0330a8a79e16
      imagePort: 80,
Started TaskList: Pushing Meteor App
[35.186.186.191] - Pushing Meteor App Bundle to the Server
Failed to copy file  { [Error: Permission denied] code: 3, lang: '' }
Retrying in 3 seconds`

@zodern there is any other config that needs to be placed.
